pkg/ingester/client/client.go

```diff
@@ -72,6 +88,32 @@ func (c *closableHealthAndIngesterClient) Push(ctx context.Context, in *cortexpb
 	})
 }
 
+func (c *closableHealthAndIngesterClient) PushStreamConnection(ctx context.Context, in *cortexpb.WriteRequest, opts ...grpc.CallOption) (*cortexpb.WriteResponse, error) {
+	return c.handlePushRequest(func() (*cortexpb.WriteResponse, error) {
+		tenantID, err := tenant.TenantID(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		streamReq := &cortexpb.StreamWriteRequest{
+			TenantID: tenantID,
+			Request:  in,
+		}
+
+		reqCtx, reqCancel := context.WithCancel(ctx)
+		defer reqCancel()
+
+		job := &streamWriteJob{
+			req:    streamReq,
+			ctx:    reqCtx,
+			cancel: reqCancel,
+		}
+		c.streamPushChan <- job
```

Comment:
I am not sure if it makes any difference here. Maybe we can add a check for `c.streamCtx` to see if ctx is already canceled before pushing to the channel. Since those jobs are going to be drained/not processed anyway.